rust/src/shards/gui/layouts/table.rs

```diff
@@ -0,0 +1,421 @@
+        }
+        Ok(self.columns.set_param(value))
+      }
+      2 => Ok(self.striped.set_param(value)),
+      3 => Ok(self.resizable.set_param(value)),
+      4 => Ok(self.row_index.set_param(value)),
+      _ => Err("Invalid parameter index"),
+    }
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    match index {
+      0 => self.rows.get_param(),
+      1 => self.columns.get_param(),
+      2 => self.striped.get_param(),
+      3 => self.resizable.get_param(),
+      4 => self.row_index.get_param(),
+      _ => Var::default(),
+    }
+  }
+
+  fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
+    self.requiring.clear();
+
+    // Add UI.Parents to the list of required variables
+    util::require_parents(&mut self.requiring, &self.parents);
+
+    Some(&self.requiring)
+  }
+
+  fn exposedVariables(&mut self) -> Option<&ExposedTypes> {
+    self.exposing.clear();
+
+    let mut exposed = false;
+    for s in &self.shards {
+      exposed |= util::expose_contents_variables(&mut self.exposing, s);
+    }
+
+    if exposed {
+      Some(&self.exposing)
+    } else {
+      None
+    }
+  }
+
+  fn hasCompose() -> bool {
+    true
+  }
+
+  fn compose(&mut self, data: &InstanceData) -> Result<Type, &str> {
+    let mut data = *data;
+    let mut shared: ExposedTypes;
+
+    // we need to inject the row index to the inner shards
+    if self.row_index.is_variable() {
+      // clone shared into a new vector we can append things to
+      shared = data.shared.into();
+      let mut should_expose = true;
+      for var in &shared {
+        let (a, b) = unsafe {
+          (
+            CStr::from_ptr(var.name),
+            CStr::from_ptr(self.row_index.get_name()),
+          )
+        };
+        if CStr::cmp(a, b) == Ordering::Equal {
+          should_expose = false;
+          if var.exposedType.basicType != SHType_Int {
+            return Err("Table: int variable required.");
+          }
+          break;
+        }
+      }
+
+      if should_expose {
+        // expose row index
+        let index_info = ExposedInfo {
+          exposedType: common_type::int,
+          name: self.row_index.get_name(),
+          help: cstr!("The row index.").into(),
+          isMutable: false,
+          isProtected: false,
+          isTableEntry: false,
+          global: false,
+          isPushTable: false,
+        };
+        shared.push(index_info);
+        // update shared
+        data.shared = (&shared).into();
+      }
+    }
+
+    let input_type = data.inputType;
+    let slice = unsafe {
+      let ptr = input_type.details.seqTypes.elements;
+      std::slice::from_raw_parts(ptr, input_type.details.seqTypes.len as usize)
+    };
+    let element_type = if !slice.is_empty() {
+      slice[0]
+    } else {
+      common_type::any
+    };
+
+    for s in &mut self.shards {
+      data.inputType = element_type;
+      s.compose(&data)?;
+    }
+
+    // Always passthrough the input
+    Ok(input_type)
+  }
```

Comment:
Shouldn't the header_shards be composed too?

Reply:
They should. I just forgot 😉